ir/instr.cpp

```diff
@@ -1558,8 +1577,17 @@ StateValue ConversionOp::toSMT(State &s) const {
     };
     break;
   case Trunc:
-    fn = [](auto &&val, auto &to_type) -> StateValue {
-      return {val.trunc(to_type.bits()), true};
+    fn = [this](auto &&val, auto &to_type) -> StateValue {
+      AndExpr non_poison;
+      unsigned orig_bits = val.bits();
+      unsigned trunc_bits = to_type.bits();
+      if (flags & NUW)
+        non_poison.add(val.trunc(trunc_bits).zext(orig_bits - trunc_bits) ==
+                       val);
+      if (flags & NSW)
+        non_poison.add(val.trunc(trunc_bits).sext(orig_bits - trunc_bits) ==
+                       val);
+      return {val.trunc(trunc_bits), non_poison()};
```

Comment:
I suggest don't computing trunc() 3x times by saving this in a tmp var.
These functions do pre-processing and can be expensive.

The rest of the patch is great, thanks!

Reply:
Definitely, thanks for the quick review! :)